sdk/storage/Azure.Storage.Queues/src/QueueClientSideDecryptor.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Storage.Cryptography;
+using Azure.Storage.Queues.Models;
+using Azure.Storage.Queues.Specialized;
+using Azure.Storage.Queues.Specialized.Models;
+
+namespace Azure.Storage.Queues
+{
+    internal class QueueClientSideDecryptor
+    {
+        private readonly ClientSideDecryptor _decryptor;
+        private readonly IClientSideDecryptionFailureListener _listener;
+
+        public QueueClientSideDecryptor(ClientSideDecryptor decryptor, IClientSideDecryptionFailureListener listener)
+        {
+            _decryptor = decryptor;
+            _listener = listener;
+        }
+
+        public async Task<QueueMessage[]> ClientSideDecryptMessagesInternal(QueueMessage[] messages, bool async, CancellationToken cancellationToken)
+        {
+            var filteredMessages = new List<QueueMessage>();
+            foreach (var message in messages)
+            {
+                try
+                {
+                    message.MessageText = await ClientSideDecryptInternal(message.MessageText, async, cancellationToken).ConfigureAwait(false);
+                    filteredMessages.Add(message);
+                }
+                catch (Exception e) when (_listener != default)
+                {
+                    if (async)
+                    {
+                        await _listener.OnFailureAsync(message, e).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        _listener.OnFailure(message, e);
+                    }
+                }
+            }
+            return filteredMessages.ToArray();
```

Comment:
So you have to implement a listener if you don't want to wait for the visiblity timeout on messages you can't read?  This design feels tricky to use correctly.

Reply:
The alternatives as far as we've explored are to
1. Return the serialized encrypted message schema and let the user write the code to discover they failed to decrypt.
2. Silently filter out messages we couldn't decrypt.

I'm not happy with either of those. The first is undue burden on the customer, whereas through the listener we do the determination for them and they handle it how they wish. The second feels dangerous in that queue messages can just time out with no one noticing.

I suppose an option 3 could exist of overriding the failed messages to throw upon accessing the message text and just always return all the messages, but I'm not sure whether that's useful.